cdk/src/open-data-platform/data-plane/data-import/write-demographic-data.handler.ts

```diff
@@ -1,149 +1,211 @@
-          results.push(row);
-        }
-        count += 1;
-        fileStream.resume();
-      })
-      .on('error', (error: Error) => {
-        reject(error);
-      })
-      .on('end', () => {
-        console.log('Parsed ' + results.length + ' rows.');
-        resolve(results);
-      });
+
+    let pipeline = chain([
+      fileStream,
+      parser(),
+      pick({ filter: 'features' }),
+      streamArray(),
+      new Batch({ batchSize: BATCH_SIZE }),
+    ]);
+
+    const endIndex = startIndex + numberOfRowsToWrite;
+    try {
+      pipeline
+        .on('data', (rows: any[]) => {
+          // Stop reading stream if this would exceed number of rows to write.
+          if (numberRowsParsed + rows.length > endIndex) {
+            pipeline.destroy();
+          }
+
+          let tableRows: SqlParametersList[] = [];
+          const shouldWriteRow = numberRowsParsed >= startIndex && numberRowsParsed <= endIndex;
+
+          if (shouldWriteRow) {
+            tableRows = rows.map(getTableRowFromRow);
+          }
+
+          promises.push(executeBatchOfRows(rdsDataService, tableRows));
+          numberRowsParsed += rows.length;
+        })
+        .on('error', (error: Error) => {
+          reject(error);
+        })
+        // Gets called by pipeline.destroy()
+        .on('close', async (_: Error) => {
+          await handleEndOfFilestream(promises);
+          resolve(numberRowsParsed);
+        })
+        .on('end', async () => {
+          await handleEndOfFilestream(promises);
+          resolve(numberRowsParsed);
+        });
+    } catch (error) {
+      reject(error);
+    }
   });
 }
 
+/**
+ * Writes the table rows to the db and adds execution to list of promises.
+ *
+ * @param rdsDataService: RDS service to connect to the db.
+ * @param tableRows: Rows to write to the db.
+ */
+function executeBatchOfRows(
+  rdsDataService: RDSDataService,
+  tableRows: SqlParametersList[],
+): Promise<BatchExecuteStatementResponse | null> {
+  let promise: Promise<BatchExecuteStatementResponse | null> = Promise.resolve(null);
+  if (tableRows.length > 0) {
+    promise = insertBatch(rdsDataService, tableRows);
+    promise.catch((_) => {
+    }); // Suppress unhandled rejection.
+  }
+  return promise;
+}
+
+/**
+ * Logs number and details of errors that occurred for all inserts.
+ * @param promises of SQL executions.
+ */
+async function handleEndOfFilestream(promises: Promise<BatchExecuteStatementResponse | null>[]) {
+  // Unlike Promise.all(), which rejects when a single promise rejects, Promise.allSettled
+  // allows all promises to finish regardless of status and aggregated the results.
+  const result = await Promise.allSettled(promises);
+  const errors = result.filter((p) => p.status == 'rejected') as PromiseRejectedResult[];
+  console.log(`Number errors during insert: ${errors.length}`);
+
+  // Log all the rejected promises to diagnose issues.
+  errors.forEach((error) => console.log(error.reason));
```

Comment:
Brett suggested using 'Insert error: ${error}' so it's searchable

Reply:
Done.